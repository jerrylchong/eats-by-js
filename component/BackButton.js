import React from 'react';
import {Image, TouchableOpacity, StyleSheet, View, Dimensions} from "react-native";

const BackButton = (props) => {
    return (
        <View style = {props.style}>
            <TouchableOpacity style = {styles.container} onPress = {props.onPress}>
                {props.white
                    ? <Image style = {styles.image} source ={require('../assets/whiteback.png')} />
                    : <Image style = {styles.image} source ={require('../assets/backbutton.png')} />
                }
            </TouchableOpacity>
        </View>
    )
}

export default BackButton

const windowWidth= Dimensions.get('window').width;

const styles = StyleSheet.create({
    container: {
        height: windowWidth * 0.1,
        width: windowWidth * 0.1,
        alignItems: 'flex-start',
        justifyContent: 'flex-start'
    },
    image: {
        height: 13,
        width: 15
    }
})