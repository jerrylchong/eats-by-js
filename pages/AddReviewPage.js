import React, { useState } from 'react';
import {
    StyleSheet,
    TouchableOpacity,
    SafeAreaView,
    TextInput,
    KeyboardAvoidingView,
    Text,
    View,
    Platform,
    ImageBackground,
    Dimensions
} from "react-native";
import {postReview} from '../helpers/apiHelpers';
import { connect } from 'react-redux';
import {mapReduxStateToProps, mapReduxDispatchToProps} from "../helpers/reduxHelpers";

function AddReviewPage(props) {
    const {route, navigation, token, user} = props;
    const [title, setTitle] = useState('');
    const [content, setContent] = useState('');
    const [rating, setRating] = useState('');
    const [date, setDate] = useState('');
    const [errors, setErrors] = useState({});
    const submit = () => {
        const review = { title,content,rating }
        const restaurant_id = route.params.restaurant_id;
        postReview(review, restaurant_id, token).then(response =>{
            if (!("error" in response) && !("errors" in response)){
                navigation.goBack();
            } else {
                "error" in response 
                    ? setErrors({authorized:"Not Authorized"})
                    : setErrors(response.errors);
            }
        });
    }
    const backHandler = () => navigation.goBack();
    return (
        <SafeAreaView style = {styles.container}>
            <ImageBackground style = {styles.background} source={require('../assets/background.png')}/>
            <Text style = {styles.header}>Add a Review</Text>
            <KeyboardAvoidingView behavior={Platform.OS == "ios" ? "padding" : "height"} style = {styles.list}>
                <TextInput
                    style={styles.input}
                    placeholder="Title"
                    onChangeText={(text) => {setTitle(text)}}
                    value={title}/>
                <TextInput
                    style={styles.input}
                    placeholder="Rating (out of 5)"
                    onChangeText={(text) => {setRating(text)}}
                    value={rating}/>
                <TextInput
                    style={styles.input}
                    placeholder="Content"
                    onChangeText={(text) => {setContent(text)}}
                    value={content}/>
                { Object.entries(errors).map(x => x[0] + " " + x[1][0]).map((x,i) => 
                <Text 
                    key={`${i}-error`}
                    style={styles.error}>{x}</Text>
                )}
            </KeyboardAvoidingView>
            <View style = {{width: '100%', height: '15%', justifyContent: 'space-between', alignItems: 'center'}}>
                <View style = {styles.buttonShadow}>
                    <TouchableOpacity style = {styles.button} onPress = {submit}>
                        <Text style = {styles.buttonText}>Submit</Text>
                    </TouchableOpacity>
                </View>
                <View style = {styles.buttonShadow}>
                    <TouchableOpacity style = {styles.button} onPress = {backHandler}>
                        <Text style = {styles.buttonText}>Back</Text>
                    </TouchableOpacity>
                </View>
            </View>
        </SafeAreaView>
    )
}

export default connect(mapReduxStateToProps,mapReduxDispatchToProps)(AddReviewPage)

const styles = StyleSheet.create({
    background: {
        position:'absolute',
        bottom: 0,
        width: '100%',
        height: Dimensions.get('window').width * 728/1668
    },
    container: {
        flex: 1,
        flexDirection: 'column',
        backgroundColor: '#fff',
        alignItems: 'center',
    },
    header: {
        fontSize: 20,
        marginTop: '10%'
    },
    list: {
        width: '80%',
        height: '60%',
        alignItems: 'center',
        marginTop: '10%'
    },
    inputHeader: {
        flexDirection: 'row',
        alignSelf: 'flex-start'
    },
    input: {
        borderBottomWidth: 1,
        borderColor: '#404040',
        fontSize: 12,
        width: '100%',
        height: '8%',
        paddingHorizontal: 5,
        marginBottom: '8%'
    },
    buttonShadow: {
        width: '25%',
        height: '30%',
        backgroundColor: '#ff6961',
        borderRadius: 20,
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 2,
        },
        shadowOpacity: 0.25,
        shadowRadius: 3.84,
        elevation: 5,
    },
    button: {
        height: '100%',
        width: '100%',
        alignItems: 'center',
        justifyContent: 'center',
        paddingHorizontal: 20,
        paddingVertical: 3,
    },
    buttonText: {
        fontSize: 12,
        color: 'white',
    },
    error: {
        fontSize: 14,
        color: '#fc8a1d',
        marginTop: '5%'
    }
})
