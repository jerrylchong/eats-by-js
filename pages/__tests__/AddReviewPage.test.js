import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, {shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import {AddReviewPage} from "../AddReviewPage";
import {ImageBackground, TouchableWithoutFeedback, Text, TextInput, TouchableOpacity} from "react-native";
import {SafeAreaProvider} from "react-native-safe-area-context";
import BackButton from "../../component/BackButton";
import { AirbnbRating } from 'react-native-ratings';

// snapshot test
test('renders correctly', () => {
    const tree = renderer.create(
        <SafeAreaProvider initialSafeAreaInsets={{ top: 0, left: 0, right: 0, bottom: 0 }}>
            <AddReviewPage
                navigation={{navigate: jest.fn()}}
            />
        </SafeAreaProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
})

Enzyme.configure({ adapter: new Adapter() });

test('test component rendering', () => {
    const wrapper = shallow(
        <SafeAreaProvider initialSafeAreaInsets={{ top: 0, left: 0, right: 0, bottom: 0 }}>
            <AddReviewPage
                navigation={{navigate: jest.fn()}} // pass mock function as goBack navigation prop
            />
        </SafeAreaProvider>
    )
    expect(wrapper.find(TouchableWithoutFeedback)).toBeTruthy(); // checks if touchable is rendered
    expect(wrapper.childAt(0).childAt(0).find(ImageBackground)).toBeTruthy(); // checks if imagebackground is rendered
    expect(wrapper.childAt(0).childAt(0).find(BackButton)).toBeTruthy(); // checks if backbutton is rendered
    expect(wrapper.childAt(0).childAt(0).find(Text)).toBeTruthy(); // checks if header text is rendered
    expect(wrapper.childAt(0).childAt(0).childAt(3).find(TextInput)).toBeTruthy(); // checks if textinput is rendered
    expect(wrapper.childAt(0).childAt(0).childAt(3).find(AirbnbRating)).toBeTruthy(); // checks if airbnbrating is rendered
    expect(wrapper.childAt(0).childAt(0).childAt(4).find(TouchableOpacity)).toBeTruthy(); // checks if submit button is rendered
    expect(wrapper.childAt(0).childAt(0).childAt(4).find(Text)).toBeTruthy(); // checks if submit button text is rendered
})