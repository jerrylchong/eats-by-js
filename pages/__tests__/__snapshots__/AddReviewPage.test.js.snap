// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "flexDirection": "column",
          "minHeight": 1334,
        },
        Object {
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Object {
          "bottom": 0,
          "height": 327.33812949640287,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <Image
        source={1}
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": 327.33812949640287,
              "width": "100%",
            },
            undefined,
          ]
        }
      />
    </View>
    <View
      style={
        Object {
          "alignSelf": "flex-start",
          "margin": "2%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
            "opacity": 1,
            "width": 75,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 13,
              "width": 15,
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "alignSelf": "center",
          "fontFamily": "Ubuntu-Medium",
          "fontSize": 20,
        }
      }
    >
      Add a Review
    </Text>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "height": "60%",
            "marginTop": "10%",
            "width": "80%",
          },
          Object {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        multiline={true}
        onChangeText={[Function]}
        placeholder="Content"
        placeholderStyle={
          Object {
            "margin": "2%",
          }
        }
        placeholderTextColor="#404040"
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "#d9d9d9",
            "borderRadius": 10,
            "color": "#404040",
            "fontFamily": "Ubuntu",
            "fontSize": 12,
            "fontWeight": "normal",
            "height": "70%",
            "marginBottom": "8%",
            "padding": "3%",
            "width": "100%",
          }
        }
        textAlignVertical="top"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "flexDirection": "column",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontWeight": "bold",
                "margin": 10,
              },
              Object {
                "color": "rgba(230, 196, 46, 1)",
                "fontSize": 25,
              },
            ]
          }
        >
          im alive
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Image
              source={1}
              style={
                Object {
                  "height": 40,
                  "margin": 3,
                  "tintColor": "#f1c40f",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "width": 40,
                }
              }
            />
          </View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Image
              source={1}
              style={
                Object {
                  "height": 40,
                  "margin": 3,
                  "tintColor": "#f1c40f",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "width": 40,
                }
              }
            />
          </View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Image
              source={1}
              style={
                Object {
                  "height": 40,
                  "margin": 3,
                  "tintColor": "#f1c40f",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "width": 40,
                }
              }
            />
          </View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Image
              source={1}
              style={
                Object {
                  "height": 40,
                  "margin": 3,
                  "tintColor": undefined,
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "width": 40,
                }
              }
            />
          </View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Image
              source={1}
              style={
                Object {
                  "height": 40,
                  "margin": 3,
                  "tintColor": undefined,
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                  "width": 40,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": "15%",
          "justifyContent": "space-evenly",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#ff6961",
            "borderRadius": 20,
            "elevation": 5,
            "height": "30%",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 3.84,
            "width": "25%",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "height": "100%",
              "justifyContent": "center",
              "opacity": 1,
              "paddingHorizontal": 20,
              "paddingVertical": 3,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontFamily": "Ubuntu",
                "fontSize": 12,
              }
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
